controllers/keyword_test.go

```diff
@@ -0,0 +1,124 @@
+	"testing"
+
+	"github.com/bxcodec/faker/v3"
+	"github.com/gin-gonic/gin"
+	"github.com/gutakk/go-google-scraper/db"
+	"github.com/gutakk/go-google-scraper/models"
+	testConfig "github.com/gutakk/go-google-scraper/tests/config"
+	testDB "github.com/gutakk/go-google-scraper/tests/db"
+	testFile "github.com/gutakk/go-google-scraper/tests/file"
+	"github.com/gutakk/go-google-scraper/tests/fixture"
+	testHttp "github.com/gutakk/go-google-scraper/tests/http"
+	"github.com/stretchr/testify/suite"
+	"golang.org/x/crypto/bcrypt"
+	"gopkg.in/go-playground/assert.v1"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+)
+
+type KeywordDbTestSuite struct {
+	suite.Suite
+	engine *gin.Engine
+	userID uint
+}
+
+func (s *KeywordDbTestSuite) SetupTest() {
+	database, _ := gorm.Open(postgres.Open(testDB.ConstructTestDsn()), &gorm.Config{})
+	db.GetDB = func() *gorm.DB {
+		return database
+	}
+
+	_ = db.GetDB().AutoMigrate(&models.User{}, &models.Keyword{})
+
+	s.engine = testConfig.GetRouter(true)
+	new(LoginController).applyRoutes(EnsureGuestUserGroup(s.engine))
+	new(KeywordController).applyRoutes(EnsureAuthenticatedUserGroup(s.engine))
+
+	email := faker.Email()
+	password := faker.Password()
+
+	hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	user := models.User{Email: email, Password: string(hashedPassword)}
+	db.GetDB().Create(&user)
+	s.userID = user.ID
+}
+
+func (s *KeywordDbTestSuite) TearDownTest() {
+	db.GetDB().Exec("DELETE FROM keywords")
+	db.GetDB().Exec("DELETE FROM users")
+}
+
+func TestKeywordDbTestSuite(t *testing.T) {
+	suite.Run(t, new(KeywordDbTestSuite))
+}
+
+func TestDisplayKeywordWithAuthenticatedUser(t *testing.T) {
+	engine := testConfig.GetRouter(true)
+	new(KeywordController).applyRoutes(EnsureAuthenticatedUserGroup(engine))
+
+	// Cookie from login API Set-Cookie header
+	headers := http.Header{}
+	cookie := "go-google-scraper=MTYwNjQ2Mjk3MXxEdi1CQkFFQ180SUFBUkFCRUFBQUlmLUNBQUVHYzNSeWFXNW5EQWtBQjNWelpYSmZhV1FFZFdsdWRBWUVBUDRFdFE9PXzl6APqAQw3gAQqlHoXMYrPpnqPFkEP8SRHJZEpl-_LDQ=="
+	headers.Set("Cookie", cookie)
+
+	response := testHttp.PerformRequest(engine, "GET", "/keyword", headers, nil)
+	p, err := ioutil.ReadAll(response.Body)
+	isKeywordPage := err == nil && strings.Index(string(p), "<title>Keyword</title>") > 0
+
+	assert.Equal(t, http.StatusOK, response.Code)
+	assert.Equal(t, true, isKeywordPage)
+}
+
+func TestDisplayKeywordWithGuestUser(t *testing.T) {
+	engine := testConfig.GetRouter(true)
+	new(KeywordController).applyRoutes(EnsureAuthenticatedUserGroup(engine))
+
+	response := testHttp.PerformRequest(engine, "GET", "/keyword", nil, nil)
+
+	assert.Equal(t, http.StatusFound, response.Code)
+	assert.Equal(t, "/login", response.Header().Get("Location"))
+}
+
+func (s *KeywordDbTestSuite) TestUploadKeywordWithAuthenticatedUserAndValidParams() {
+	headers, payload := testFile.CreateMultipartPayload("tests/fixture/adword_keywords.csv")
+	cookie := fixture.GetCookie("user_id", s.userID)
```

Comment:
I am not sure if it should be called `GetCookie` as it seems to actually generate encoded cookie data unless I am missing something in the helper code 🤔 So `fixture.GenerateCookie("user_id", s.userID)` might make more sense.

Reply:
Make more sense. Fixed in c2abe467ca7f77b75aeb578ec9b69285bc0026ab